libtock/crt0.c

```diff
@@ -63,8 +63,19 @@ void _start(void* app_start,
   // bss_size from start of memory. Also make sure that the stack starts on an
   // 8 byte boundary per section 5.2.1.2 here:
   // http://infocenter.arm.com/help/topic/com.arm.doc.ihi0042f/IHI0042F_aapcs.pdf
-  struct hdr* myhdr = (struct hdr*)app_start;
-  uint32_t stacktop = (((uint32_t)mem_start + myhdr->stack_size + 7) & 0xfffffff8);
+
+  // Before adjusting the stack, make sure to keep all variables in registers.
+  register struct hdr* myhdr = (struct hdr*)app_start;
+  register uint32_t stacktop = (((uint32_t)mem_start + myhdr->stack_size + 7) & 0xfffffff8);
+
+  {
+    register uint32_t heap_size = myhdr->got_size + myhdr->data_size + myhdr->bss_size;
+    memop(0, stacktop + heap_size);
+    memop(11, stacktop + heap_size);
+    memop(10, stacktop);
+    asm volatile ("mov sp, %[stacktop]" :: [stacktop] "r" (stacktop) : "memory");
```

Comment:
Given that `memop` under the hood will trigger a syscall, and that sycall will cause and interrupt, and the interrupt handler will store things on the stack, I think we should really move this to happen before any function calls.

Reply:
We can't move it before the the call to sbrk (the first `memop`), since that memory isn't accessible to the process yet. We _anyway_ need to ensure that the kernel gives the process enough stack to do a system call (which it does), and that's the maximum we're using here.

That was my thought, anyway.

Reply:
👍 I'd forgotten about the (fairly obvious) need for a minimum initial stack size. Is that documented anywhere (I'd have guessed here https://github.com/tock/tock/blob/master/doc/Userland.md#application-entry-point, but it's missing)?

I'll look more seriously through the docs later and add it if it's missing.